### ISSUE TITLE
With RBNO, announcing new CDC generations happens too late, causing inconsistencies 

### ISSUE THREAD

**Issue by kbr-** [scylladb/scylla#10149]
When a new CDC generation is created, it's supposed to be announced through gossip **immediately**, so that all nodes learn about it before their clocks reach the generation's timestamp. This uses the ring_delay assumption that existed since C*, i.e. that nodes learn about tokens of a bootstrapping node within 30 seconds.
This is documented (last paragraph):
```cpp
    /* Generate a new set of CDC streams and insert it into the internal distributed CDC generations table.
     * Returns the ID of this new generation.
     *
     * Should be called when starting the node for the first time (i.e., joining the ring).
     *
     * Assumes that the system_distributed_keyspace service is initialized.
     * `cluster_supports_generations_v2` must be `true` if and only if the `CDC_GENERATIONS_V2` feature is enabled.
     *
     * If `CDC_GENERATIONS_V2` is enabled, the new generation will be inserted into
     * `system_distributed_everywhere.cdc_generation_descriptions_v2` and the returned ID will be in the v2 format.
     * Otherwise the new generation will be limited in size, causing suboptimal stream distribution, it will be inserted
     * into `system_distributed.cdc_generation_descriptions` and the returned ID will be in the v1 format.
     * The second case should happen only when we create new generations in a mixed cluster.
     *
     * The caller of this function is expected to insert the ID into the gossiper as fast as possible,
     * so that other nodes learn about the generation before their clocks cross the generation's timestamp
     * (not guaranteed in the current implementation, but expected to be the common case;
     *  we assume that `ring_delay` is enough for other nodes to learn about the new generation).
     */
    future<cdc::generation_id> make_new_generation(const std::unordered_set<dht::token>& bootstrap_tokens, bool add_delay);
```

However, when RBNO is enabled, the bootstrapping node does not announce the new generation until after bootstrapping finishes, because the announce is protected with a flag that is false:
```cpp
        _cdc_gen_id = _cdc_gen_service.local().make_new_generation(_bootstrap_tokens, !is_first_node()).get0();

      if (!bootstrap_rbno) {
        // When is_repair_based_node_ops_enabled is true, the bootstrap node
        // will use node_ops_cmd to bootstrap, bootstrapping gossip status is not needed for bootstrap.
        _gossiper.add_local_application_state({
            // Order is important: both the CDC streams timestamp and tokens must be known when a node handles our status.
            { gms::application_state::TOKENS, versioned_value::tokens(_bootstrap_tokens) },
            { gms::application_state::CDC_GENERATION_ID, versioned_value::cdc_generation_id(_cdc_gen_id) },
            { gms::application_state::STATUS, versioned_value::bootstrapping(_bootstrap_tokens) },
        }).get();
```

The announce happens after bootstrapping finishes (in `set_gossip_tokens` call), which may be about a few hours too late.

This will cause writes to race with the generation announce when it finally happens, causing different coordinators to pick different stream IDs for the same partition key when writing to CDC, causing clients to miss some writes.
So basically CDC is broken in 4.6 if RBNO is used.

**Comment by kbr-** [comment]
The fix should be easy, just call
```cpp
        _gossiper.add_local_application_state({
            { gms::application_state::CDC_GENERATION_ID, versioned_value::cdc_generation_id(_cdc_gen_id) },
        }).get();
```

unconditionally after `make_new_cdc_generation`. Since 4.5 we don't need a STATUS update for nodes to notice a CDC_GENERATION_ID change (untested if it works without a STATUS update but in theory it should because there's a separate listener service).

The better question is how to catch this regression in a test. But that should also be doable, I'll think about it

**Comment by kbr-** [comment]
> But then I don't know how bootstrap works with RBNO, @asias how do other nodes learn about our tokens during RBNO bootstrap, if not through gossiper state update?

It seems they learn through direct RPC, which I kinda like, but it's sad that the RPC does not contain info about the CDC generation update :(

**Comment by asias** [comment]
We currently only enable RBNO for replace ops by default. RBNO for bootstrap is not enabled by default. 

With RBNO for bootstrap enabled, the other nodes know the tokens of the bootstrapping node by a RPC verb directly. 

auto req = node_ops_cmd_request(node_ops_cmd::bootstrap_prepare, uuid, ignore_nodes, {}, {}, bootstrap_nodes);

Since we always have the _cdc_gen_id in storage_service::bootstrap, is it enough to call unconditionally?

_gossiper.add_local_application_state({
         { gms::application_state::CDC_GENERATION_ID, versioned_value::cdc_generation_id(_cdc_gen_id) },
 }).get();

What would happen if the bootstrapping failed and the new node went away, but the new cdc id was learned by the other existing nodes. Do we need to revert something for cdc?

**Comment by kbr-** [comment]
> We currently only enable RBNO for replace ops by default. RBNO for bootstrap is not enabled by default.

Ok. For replace we don't need to change the CDC generation because no tokens change.

> What would happen if the bootstrapping failed and the new node went away, but the new cdc id was learned by the other existing nodes. Do we need to revert something for cdc?

That's fine, the new generation is good for both new and old token ring. The new generation was created so that each vnode in the new token ring has its own stream ID. Then each vnode in the old token ring also has its own stream ID because the vnodes in the new token ring are smaller (contained in the vnodes of the old token ring). So even if we stay with the old token ring we're good.

**Comment by kbr-** [comment]
Backported to 4.6
